docs: add page on sync modes (#25634)
ethereum/go-ethereum: docs/_interface/sync-modes.md

---
title: Sync-modes
sort-key: L
---

Syncing is the process by which Geth catches up to the latest Ethereum block and current global state. 
There are several ways to sync a Geth node that differ in their speed, storage requirements and trust 
assumptions. This page outlines three sync configurations for full nodes and one for light nodes.

## Full nodes

There are two types of full node that use different mechanisms to sync up to the head of the chain:

### Snap (default)

> A snap sync'd node holds the most recent 128 blocks in memory, so transactions in that range are always 

Reviewer: the most recent 128 block states in memory. And transactions are always accessible I think :)

> A snap sync'd node holds the most recent 128 blocks in memory, so transactions in that range are always 

Author: ```suggestion
A snap sync'd node holds the most recent 128 block states in memory, so transactions in that range are always quickly 
```

accessible. However, snap-sync only starts processing from a relatively recent block (as opposed to genesis 

> for a full node). Between the initial sync block and the 128 most recent blocks, the node stores occasional 

Reviewer: The periodically flush is actually the in-memory state pruning mechanism. It's not relevant with snap sync at all. In another word, no matter it's full sync'd, snap sync'd or even legacy fast sync'd, they will also periodically flush state into disk and prune all others unless `gcmode = archive`.

I will vote to explain a bit more snap-sync specified maybe.

Reviewer: for example, we can explain the workflow of the snap sync, it can be divided into different stages
- backward header sync(with some security consideration, etc)
- backfill corresponding block bodies/receipts in between
- at the same time, retrieve the state data

Reviewer: In state sync, it can also be divided to two stages:
- sync stage: retrieve raw state data(accounts, storage slots) with range proof attached
- healing stage: fulfill the missing trie nodes because of the syncing target moving

progressing at the same time and invalidating some of the regenerated state data. This means it is also necessary 
to have a 'healing' phase where errors in the state are fixed. It is not possible to monitor the progress of 
the state heal because the extent of the errors cannot be known until the current state has already been regenerated. 
The healing has to outpace the growth of the blockchain, otherwise the node will never catch up to the current state. 
There are some hardware factors that determine the speed of the state healing (speed of disk read/write and internet 
connection) and also the total gas used in each block (more gas means more changes to the state that have to be 
handled).

**Note** Snap sync is the default behaviour, so if the `--syncmode` value is not passed to Geth at startup, 
Geth will use snap sync. A node that is started using `snap` will switch to block-by-block sync once it has 
caught up to the head of the chain.

### Full

A full sync generates the current state by executing every block starting from the genesis block. A full sync 
indendently verifies proof-of-work and block provenance as well as all state transitions by re-executing the 

> transactions in the entire historical sequence of blocks. Only the most recent 128 blocks are stored in a full 

Reviewer: Only the most recent 128 blocks state are kept in a full node, - the state of older blocks are pruned periodically

> transactions in the entire historical sequence of blocks. Only the most recent 128 blocks are stored in a full 

Author: ```suggestion
transactions in the entire historical sequence of blocks. Only the most recent 128 block states are stored in a full 
```

> node - older blocks are pruned periodically and represented as a series of checkpoints from which any previous 

Author: ```suggestion
node - older block states are pruned periodically and represented as a series of checkpoints from which any previous 
```

garbage collection so that old data is never deleted: `geth --syncmode full --gcmode archive`. 

It is also possible to create a partial/recent archive node where the node was synced using `snap` but the state 
is never pruned. This creates an archive node that saves all state data from the point that the node first syncs. 
This is configured by starting Geth with `--syncmode snap gcmode archive`.

## Light nodes

A light node syncs very quickly and stores the bare minimum of blockchain data. Light nodes only process block
headers, not entire blocks. This greatly reduces the computation time, storage and bandwidth required relative to a 
full node. This means light nodes are suitable for resource-constrained devices and can catch up to the head of the
chain much faster when they are new or have been offline for a while. The trade-off is that light nodes rely heavily 
on data served by altruistic full nodes. A light client can be used to query data from Ethereum and submit transactions, 
acting as a locally-hosted Ethereum wallet. However, because they don't keep local copies of the Ethereum state, light 

> nodes can't validate blocks in the same way as full nodes - they have to trust that the data they receive is honest. 

Reviewer: Light client does sacrifice some security guarantee, but it's not totally trust the received data. Instead, full node needs to attach a proof in the responded packet and light client can verify it against the local header chain.

> nodes can't validate blocks in the same way as full nodes - they have to trust that the data they receive is honest. 

Author: ```suggestion
nodes can't validate blocks in the same way as full nodes - they receive a proof from the full node and verify it against their local header chain. 
```

from its peers is correct during the downloading phase but then retroactively verifies each downloaded block. Nodes 
are not allowed to attest or propose blocks while they are still 'optimistic' because they can't yet guarantee their 
view of the ehad of the chain is correct.

Read more in the [optimistic sync specs](https://github.com/ethereum/consensus-specs/blob/dev/sync/optimistic.md).

### Checkpoint sync

Alternatively, the consensus client can grab a checkpoint from a trusted source which provides a target state to sync 
up to, before switching to full sync and verifying each block in turn. In this mode, the node trusts that the checkpoint 
is correct. There are many possible sources for this checkpoint - the gold standard would be to get it out-of-band 
from another trusted friend, but it could also come from block explorers or public APIs/web apps.

**Note** it is not currently possible to use a Geth light node as an execution client on proof-of-stake Ethereum.

## Summary

There are several ways to sync a Geth node. The default is to use snap sync to create a full node. This verifies all 

> blocks starting at a recent checkpoint. A trust-minimized alternative is full-sync, which verifies every block since genesis. 

Reviewer: recent checkpoint feels weird. Actually there is no checkpoint notion in snap sync. We just basically choose a recent block as the syncing target. The target won't be chosen as the current head since it might be reorged. Instead we usually choose `current.head - 64` which is new enough but still has enough block confirmations on top.

> These modes prune the blockchain data older than 128 blocks, keeping only checkpoints that enable on-request regeneration of 

Reviewer: we don't prune blockchain data, we only prune state data older than 128 blocks. And to be fair, it's not the actual pruning, it's in-memory garbage collection by dropping the old state out of the memory.

---
title: Sync-modes
sort-key: L
---

Syncing is the process by which Geth catches up to the latest Ethereum block and current global state. 
There are several ways to sync a Geth node that differ in their speed, storage requirements and trust 
assumptions. This page outlines three sync configurations for full nodes and one for light nodes.

## Full nodes

There are two types of full node that use different mechanisms to sync up to the head of the chain:

### Snap (default)

A snap sync'd node holds the most recent 128 block states in memory, so transactions in that range are always quickly 
accessible. However, snap-sync only starts processing from a relatively recent block (as opposed to genesis 
for a full node). Between the initial sync block and the 128 most recent blocks, the node stores occasional 
checkpoints that can be used to rebuild the state on-the-fly. This means transactions can be traced back as 
far as the block that was used for the initial sync. Tracing a single transaction requires reexecuting all 
preceding transactions in the same block **and** all preceding blocks until the previous stored snapshot. 

> Snap-sync'd nodes are therefore functionally equal to full nodes, but the initial synchronization required 

Reviewer 2: A note about terminology.. A node is either a full node or a light node. 
Therefore, a snap-synced node is a full node. 
An `syncmode=full` is also a fulll node. 
So it' snot "functionally equal to full nodes". They are full nodes.

Author: ```suggestion
Snap-sync'd nodes are therefore full nodes, with the only difference being the initial synchronization required 
```

a checkpoint block to sync from instead of independently verifying the chain all the way from genesis. 
Snap sync then only verifies the proof-of-work and ancestor-child block progression and assumes that the 
state transitions are correct rather than re-executing the transactions in each block to verify the state 

> changes. Snap sync is much faster than full sync. To start a node with snap sync pass `--syncmode snap` at 

Reviewer 2: The term "full sync" is IMO confusing, because it does not align with the term "full node". I prefer to use ".. than block-by-block sync", or " .. than `syncmode=full`".

Author: ```suggestion
changes. Snap sync is much faster than block-by-block sync. To start a node with snap sync pass `--syncmode snap` at 
```

startup.

Snap sync starts by downloading the headers for a chunk of blocks. Once the headers have been verified, the block

> bodies and receipts for those blocks are downloaded. Next, state sync begins. In state-sync, Geth first downloads the 

Reviewer 2: The process is more like 
``` 
                 /-- block bodies, receipts ---> 
headers ------->
                 \-- state ----> state heal -->

Author: ```suggestion
bodies and receipts for those blocks are downloaded. In parallel, Geth also sync begins state-sync. In state-sync, Geth first downloads the 
```

> bodies and receipts for those blockjs are downloaded. Next, state sync begins. In state-sync, Geth first downloads the 

Author: ```suggestion
bodies and receipts for those blocks are downloaded. Next, state sync begins. In state-sync, Geth first downloads the 
```

then regenerated locally. The state download is the part of the snap-sync that takes the most time to complete 
and the progress can be monitored using the ETA values in the log messages. However, the blockchain is also 
progressing at the same time and invalidating some of the regenerated state data. This means it is also necessary 
to have a 'healing' phase where errors in the state are fixed. It is not possible to monitor the progress of 
the state heal because the extent of the errors cannot be known until the current state has already been regenerated. 
The healing has to outpace the growth of the blockchain, otherwise the node will never catch up to the current state. 
There are some hardware factors that determine the speed of the state healing (speed of disk read/write and internet 
connection) and also the total gas used in each block (more gas means more changes to the state that have to be 
handled).

To summarize, snap sync progresses in the following sequence:
- download and verify headers
- download block bodies and receipts
- download raw state data
- regenerate state trie

> - heal state trie to account for newly arriving data

Reviewer 2: Well, step 2 is parallel to (3-4-5). I'm not sure "regenerate state trie" should be it's own step, it's more of an internal/implicit part of downloading state.

Author: ```suggestion
To summarize, snap sync progresses in the following sequence:
- download and verify headers
- download block bodies and receipts.In parallel, download raw state data and build state trie
- heal state trie to account for newly arriving data
```

node - older block states are pruned periodically and represented as a series of checkpoints from which any previous 
state can be regenerated on request. 128 blocks is about 25.6 minutes of history with a block time of 12 seconds. 
To create a full node pass `--syncmode full` at startup.

## Archive nodes

An archive node is a node that retains all historical data right back to genesis. There is no need to regenerate 
any data from checkpoints because all data is directly available in the node's own storage. Archive nodes are 
therefore ideal for making fast queries about historical states. At the time of writing (September 2022) a full 
archive node that stores all data since genesis occupies nearly 12 TB of disk space (keep up with the current 
size on [Etherscan](https://etherscan.io/chartsync/chainarchive)). Archive nodes are created by configuring Geth's 
garbage collection so that old data is never deleted: `geth --syncmode full --gcmode archive`. 

It is also possible to create a partial/recent archive node where the node was synced using `snap` but the state 
is never pruned. This creates an archive node that saves all state data from the point that the node first syncs. 

> This is configured by starting Geth with `--syncmode snap gcmode archive`.

Reviewer 2: ```suggestion
This is configured by starting Geth with `--syncmode snap --gcmode archive`.
```

---
title: Sync modes
sort_key: L
---

Syncing is the process by which Geth catches up to the latest Ethereum block and current global state. There are several ways to sync a Geth node that differ in their speed, storage requirements and trust assumptions. This page outlines three sync configurations for full nodes and one for light nodes.

## Full nodes

> There are three types of full node with different sync modes:

Author: hmm ok, i see, thanks.
